### PR TITLE
feat: apply the rest of Optimism v1.0.3 to Kroma

### PR DESCRIPTION
**Pull request by seolaoh**
# Changelog

## Remarkable Changes
- Make changes compatible with updated geth v1.11.5
  - https://github.com/ethereum-optimism/optimism/pull/5225
- `L2BlockTime` in e2e test: 2s to 1s
  - https://github.com/ethereum-optimism/optimism/pull/5374
- Place engine api related codes in e2e folder that was separated into [op-program](https://github.com/ethereum-optimism/optimism/pull/5297/files) folder in Optimism
    - https://github.com/ethereum-optimism/optimism/pull/5335
    - https://github.com/ethereum-optimism/optimism/pull/5350
- Unify terminology in `e2e/system_test.go`
    - `l2Syncer` → `l2Sync`
    - `receiptSeq` → `receiptProp`
    - `receiptVerif` → `receiptSync`
- Add L2 contracts deploy scripts
  - https://github.com/ethereum-optimism/optimism/pull/5314
- Upgrade `forge`, `forge-std`
  - https://github.com/ethereum-optimism/optimism/pull/5227

## op-e2e
- https://github.com/ethereum-optimism/optimism/pull/5334
- https://github.com/ethereum-optimism/optimism/pull/5225
- https://github.com/ethereum-optimism/optimism/pull/5105
- https://github.com/ethereum-optimism/optimism/pull/5179
- https://github.com/ethereum-optimism/optimism/pull/5180
- https://github.com/ethereum-optimism/optimism/pull/5230
- https://github.com/ethereum-optimism/optimism/pull/5189
- https://github.com/ethereum-optimism/optimism/pull/5246
- https://github.com/ethereum-optimism/optimism/pull/5258
- https://github.com/ethereum-optimism/optimism/pull/5374
- https://github.com/ethereum-optimism/optimism/pull/5283
- https://github.com/ethereum-optimism/optimism/pull/5306
- https://github.com/ethereum-optimism/optimism/pull/5369
- https://github.com/ethereum-optimism/optimism/pull/5255
- https://github.com/ethereum-optimism/optimism/pull/5335
- https://github.com/ethereum-optimism/optimism/pull/5350
- https://github.com/ethereum-optimism/optimism/pull/5361

## op-chain-ops
- https://github.com/ethereum-optimism/optimism/pull/5365

## packages/common-ts
- https://github.com/ethereum-optimism/optimism/pull/5220
- https://github.com/ethereum-optimism/optimism/pull/5126

## packages/contracts
- https://github.com/ethereum-optimism/optimism/pull/5296
- https://github.com/ethereum-optimism/optimism/pull/5227
- https://github.com/ethereum-optimism/optimism/pull/5250
- https://github.com/ethereum-optimism/optimism/pull/5276
- https://github.com/ethereum-optimism/optimism/pull/5252
- https://github.com/ethereum-optimism/optimism/pull/5312
- https://github.com/ethereum-optimism/optimism/pull/5233
- https://github.com/ethereum-optimism/optimism/pull/5194
- https://github.com/ethereum-optimism/optimism/pull/5314
- https://github.com/ethereum-optimism/optimism/pull/5279
- https://github.com/ethereum-optimism/optimism/pull/5294
- https://github.com/ethereum-optimism/optimism/pull/5356

## docs
- https://github.com/ethereum-optimism/optimism/pull/5229
- https://github.com/ethereum-optimism/optimism/pull/5261
- https://github.com/ethereum-optimism/optimism/pull/5401
- https://github.com/ethereum-optimism/optimism/pull/5299

### CODE REVIEW

**Comment by chokobole** [comment]
Please update bindings!

```shell
> cd bindings
> make more
> git diff
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   bindings/kromaportal_more.go
        modified:   bindings/l1block_more.go
        modified:   bindings/systemconfig_more.go
```

File: packages/contracts/contracts/echidna/FuzzKromaPortal.sol [hunk 1]
```diff
@@ -3,17 +3,32 @@ pragma solidity 0.8.15;
 import { KromaPortal } from "../L1/KromaPortal.sol";
 import { L2OutputOracle } from "../L1/L2OutputOracle.sol";
 import { ZKMerkleTrie } from "../L1/ZKMerkleTrie.sol";
+import { SystemConfig } from "../L1/SystemConfig.sol";
+import { ResourceMetering } from "../L1/ResourceMetering.sol";
 import { AddressAliasHelper } from "../vendor/AddressAliasHelper.sol";
+import { Constants } from "../libraries/Constants.sol";
```
Comment (chokobole) [inline packages/contracts/contracts/echidna/FuzzKromaPortal.sol:9]: Please sort the import statements.

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 2]
```diff
@@ -410,7 +432,8 @@ contract ERC721Bridge_Initializer is Messenger_Initializer {
 
 contract Colosseum_Initializer is L2OutputOracle_Initializer {
     uint256 immutable CHAIN_ID = 901;
-    bytes32 immutable DUMMY_HASH = hex"8e556cf0e9ed5d6b6ad79247cddc30112cfee4a207fb13903eb834b447aebae9";
+    bytes32 immutable DUMMY_HASH =
+        hex"8e556cf0e9ed5d6b6ad79247cddc30112cfee4a207fb13903eb834b447aebae9";
```
Comment (chokobole) [inline packages/contracts/contracts/test/CommonTest.t.sol:436]: Let's apply format rule later at once!

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 3]
```diff
@@ -624,15 +647,9 @@ contract FFIInterface is Test {
...
-        returns (
-            bytes32,
-            bytes memory,
-            bytes memory,
-            bytes[] memory
-        )
-    {
+    function getMerkleTrieFuzzCase(
+        string memory variant
+    ) external returns (bytes32, bytes memory, bytes memory, bytes[] memory) {
```
Comment (chokobole) [inline packages/contracts/contracts/test/CommonTest.t.sol:652]: Let's apply format rule later at once!

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 4]
```diff
@@ -18,6 +19,8 @@ import { L1StandardBridge } from "../L1/L1StandardBridge.sol";
 import { L2OutputOracle } from "../L1/L2OutputOracle.sol";
 import { ZKMerkleTrie } from "../L1/ZKMerkleTrie.sol";
 import { ZKVerifier } from "../L1/ZKVerifier.sol";
+import { SystemConfig } from "../L1/SystemConfig.sol";
+import { ResourceMetering } from "../L1/ResourceMetering.sol";
 import { L2CrossDomainMessenger } from "../L2/L2CrossDomainMessenger.sol";
 import { L2ERC721Bridge } from "../L2/L2ERC721Bridge.sol";
 import { L2ToL1MessagePasser } from "../L2/L2ToL1MessagePasser.sol";
```
Comment (chokobole) [inline packages/contracts/contracts/test/CommonTest.t.sol:26]: Please sort the import statements

File: packages/contracts/deploy/106-ValidatorRewardVault.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { assertContractVariable, deploy } from '../src/deploy-utils'
+
+const deployFn: DeployFunction = async (hre) => {
+  const l1 = hre.network.companionNetworks['l1']
+  const deployConfig = hre.getDeployConfig(l1)
+
+  const validatorRewardVaultRecipient =
+    deployConfig.validatorRewardVaultRecipient
+  if (validatorRewardVaultRecipient === ethers.constants.AddressZero) {
+    throw new Error(`ValidatorRewardVault RECIPIENT undefined`)
```
Comment (chokobole) [inline packages/contracts/deploy/106-ValidatorRewardVault.ts:15]:
```suggestion
    throw new Error(`ValidatorRewardVault RECIPIENT zero address`)
```

File: packages/contracts/deploy/107-ProtocolVault.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import { assertContractVariable, deploy } from '../src/deploy-utils'
+
+const deployFn: DeployFunction = async (hre) => {
+  const l1 = hre.network.companionNetworks['l1']
+  const deployConfig = hre.getDeployConfig(l1)
+
+  const protocolVaultRecipient = deployConfig.protocolVaultRecipient
+  if (protocolVaultRecipient === ethers.constants.AddressZero) {
+    throw new Error('ProtocolVault RECIPIENT undefined')
```
Comment (chokobole) [inline packages/contracts/deploy/107-ProtocolVault.ts:14]:
```suggestion
    throw new Error(`ProtocolVault RECIPIENT zero address`)
```

File: packages/contracts/deploy/108-ProposerRewardVault.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+import { assertContractVariable, deploy } from '../src/deploy-utils'
+
+const deployFn: DeployFunction = async (hre) => {
+  const l1 = hre.network.companionNetworks['l1']
+  const deployConfig = hre.getDeployConfig(l1)
+
+  const ProposerRewardVaultRecipient = deployConfig.ProposerRewardVaultRecipient
+  if (ProposerRewardVaultRecipient === ethers.constants.AddressZero) {
+    throw new Error('ProposerRewardVault RECIPIENT undefined')
```
Comment (chokobole) [inline packages/contracts/deploy/108-ProposerRewardVault.ts:14]:
```suggestion
    throw new Error(`ProposerRewardVault RECIPIENT zero address`)
```

File: packages/contracts/contracts/L1/KromaPortal.sol [hunk 8]
```diff
@@ -289,10 +312,9 @@ contract KromaPortal is Initializable, ResourceMetering, Semver {
      *
      * @param _tx Withdrawal transaction to finalize.
      */
-    function finalizeWithdrawalTransaction(Types.WithdrawalTransaction memory _tx)
-        external
-        whenNotPaused
-    {
+    function finalizeWithdrawalTransaction(
+        Types.WithdrawalTransaction memory _tx
+    ) external whenNotPaused {
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/KromaPortal.sol:317]: I see! Should I revert the changes related to formatting?

**Comment by chokobole** [comment]
> Please update bindings!
> 
> ```shell
> > cd bindings
> > make more
> > git diff
> Changes not staged for commit:
>   (use "git add <file>..." to update what will be committed)
>   (use "git restore <file>..." to discard changes in working directory)
>         modified:   bindings/kromaportal_more.go
>         modified:   bindings/l1block_more.go
>         modified:   bindings/systemconfig_more.go
> ```

When I did like below, it gives outputs like above.

```shell
# at packages/contracts
> npx hardhat compile
# at bindings
> make more
```

Strangely, if I tried after removing `artifacts`, it gives the same output as this PR.

**Comment by seolaoh** [comment]
> > Please update bindings!
> > ```shell
> > > cd bindings
> > > make more
> > > git diff
> > Changes not staged for commit:
> >   (use "git add <file>..." to update what will be committed)
> >   (use "git restore <file>..." to discard changes in working directory)
> >         modified:   bindings/kromaportal_more.go
> >         modified:   bindings/l1block_more.go
> >         modified:   bindings/systemconfig_more.go
> > ```
> 
> When I did like below, it gives outputs like above.
> 
> ```shell
> # at packages/contracts
> > npx hardhat compile
> # at bindings
> > make more
> ```
> 
> Strangely, if I tried after removing `artifacts`, it gives the same output as this PR.

I think you should delete your ./packages/contracts/cache folder, that is created before by `npx hardhat compile`.
So you should run like below to regenerate bindings.
```shell
# at ./packages/contracts
> yarn clean && npx hardhat compile
# at ./bindings
> make more
```
or just simply
```shell
# at ./
> make bindings
```

**Comment by seolaoh** [comment]
I reflected your comments and added a few new commits.

File: specs/withdrawals.md [hunk 9]
```diff
@@ -237,3 +216,6 @@ contracts if desired.
 | `FINALIZATION_PERIOD` | `604_800` | seconds |
 
 This `FINALIZATION_PERIOD` value is equivalent to 7 days.
+
+[`WithdrawalTransaction` type]: https://github.com/kroma-network/kroma/blob/dev/packages/contracts/contracts/libraries/Types.sol#L131-L138
+[`OutputRootProof` type]: https://github.com/kroma-network/kroma/blob/dev/packages/contracts/contracts/libraries/Types.sol#L34-L40
```
Comment (chokobole) [inline specs/withdrawals.md:221]: Would it be better to maintain the link with commit sha instead of the branch name, since branches can easily change?

**Review by chokobole — APPROVED** [review]
LGTM

**Pull request merged**